### DISCUSSION TITLE
Removing OBX segments using a 'white list'

### DISCUSSION THREAD

**Discussion by Pensive-Potato** [nextgenhealthcare/connect#4891] [Q&A] [answered]
Good Day,

I'm using Mirth v. 3.11

I'm working on a transformer step to edit an incoming Hl7 message on a destination where I need to:
- change obx-3-1 to a different value
- change obx-6-1 to a different value
- remove any 'empty' obx lines that contain no obx-1-1  number

So far, so good; the above 3 actions are working.  As a last task, I need to remove any unwanted obx segments, then renumber the segments afterwards and it is here that things aren't working.  Here is the code I've created in the destination transformer:

```js
var counter = 0;
var setId = 1;

		//Remove 'empty' OBX lines
		 
	for each(obx in msg..OBX) {
		if(obx['OBX.1']['OBX.1.1'].toString() == '') 
			{ delete msg.OBX[counter]; }		
	
		// White List lookup table to exclude unwanted parameters from OBX output
			
const whiteList = ["RESP","SYS","DIA","MAP","SPO2","TEMPERATURE","HR","Bewustzijn","Zuurstoftoediening","PAIN","MEWS"];
var whiteList_org = obx['OBX.3']['OBX.3.1'].toString();
var whiteListCheck = whiteList.includes(whiteList_org);

	if (whiteListCheck == false) {
		logger.info("Parameter not found in white list, will be excluded from ORU output ");
		delete msg['OBX'][counter];} 
	else { obx['OBX.1']['OBX.1.1'] = setId;
		setId++; }

		//Replace OBX 3.1 Observation Code value with that from hash table in the preprocessor script
		
var obsId_org = obx['OBX.3']['OBX.3.1'].toString();
var obsId = $('ObsCodes').get(obsId_org);

	if ( obsId != null ) 
		{ obx['OBX.3']['OBX.3.1'] = obsId; }
		
		//Replace OBX 6.1 Unit Of Measure value with that from hash table in the preprocessor script
		
var UOMId_org = obx['OBX.6']['OBX.6.1'].toString();
var UOMId = $('UOMCodes').get(UOMId_org);

	if ( UOMId != null ) 
		{ obx['OBX.6']['OBX.6.1'] = UOMId; }
	
counter++;
}
```
Here is the deploy script:
```js
// Translation Of original OBX 3.1 to new observation parameter names
var map = new java.util.HashMap();
map.put("SYS", "0002-4A05^NiBDs^MDIL");
map.put("DIA", "0002-4A06^NiBDd^MDIL");
map.put("MAP", "0002-4A07^NiBDm^MDIL");
map.put("TEMPERATURE", "0002-FB4B^iTtymp^MDIL");
map.put("RESP", "0002-F828^SpRF^MDIL");
map.put("SPO2", "0002-4BB8^SpO2^MDIL");
map.put("PAIN", "0002-F98C^Pijn^MDIL");
map.put("HR", "0002-4822^Pols^MDIL");
map.put("Zuurstoftoediening", "0002-FB58^Zuurstof^MDIL");
map.put("Bewustzijn", "0002-F989^Bewustzijn^MDIL");
map.put("MEWS", "0002-F98A^NEWS^MDIL");
globalMap.put('ObsCodes', map);


// Translation Of original OBX 6.1 to new unit of measure values
var map = new java.util.HashMap();
map.put("Mmhg", "0004-0F20^mmHg^MDIL");
map.put("Celsius", "0004-17A0^°C^MDIL");
map.put("BPM", "0004-0AE0^rpm^MDIL");
map.put("BeatsPerMinute", "0004-0AA0^spm^MDIL");
map.put("%", "0004-0220^%^MDIL");
globalMap.put('UOMCodes', map);
```

Here is the Hl7 message that comes in:

MSH|^~\&|Connex|Welch Allyn|HIS||20211209124449||ORU^R01|014AD5C5-5CE6-444C-B010-78055F5FB6E1|P|2.5
PID|||0012345678||TYSON^MIKE
PV1||I|||||||||||||||||0087654321
OBR||||S|||20211209124436|||123|||||||||||||||F
OBX|1|NM|HT||183.0|Cm||N|||F|||20211209124436|100024000517|123||Manual
OBX|2|NM|BMI||23.0|||null|||F|||20211209124436|100024000517|123||Manual
OBX|3|NM|WT||77.0|Kg||N|||F|||20211209124436|100024000517|123||Manual
OBX|4|NM|PAIN||5|||N|||F|||20211209124436|100024000517|123||Manual
OBX|5|ST|Mode||None||||||F|||20211209124436|100024000517|123||Manual
OBX|6|NM|RESP||14|/min||N|||F|||20211209124436|100024000517|123||Manual
OBX|7|NM|SYS||120|mmHg||N|||F|||20211209124436|100024000517|123||Manual
OBX|8|NM|DIA||80|mmHg||N|||F|||20211209124436|100024000517|123||Manual
OBX|9|NM|MAP||0.0|mmHg||null|||F|||20211209124436|100024000517|123||Manual
OBX|10|NM|SPO2||99|%||N|||F|||20211209124436|100024000517|123||Manual
OBX|11|ST|TEMPSITE||None||||||F|||20211209124436|100024000517|123||Manual
OBX|12|NM|TEMPERATURE||36.9|°C||N|||F|||20211209124436|100024000517|123||Manual
OBX|13|NM|HR||65|/min||N|||F|||20211209124436|100024000517|123||Manual
OBX|14|ST|Bewustzijn||Pijn reactie|||null|||F|||20211209124436|100024000517|123
OBX|15|ST|Zuurstoftoediening||Ja|||null|||F|||20211209124436|100024000517|123
OBX|16|NM|SpO2::SaturationScore11||0||||||F|||20211209124436|100024000517|123|MEWS
OBX|17|ST|SpO2::SaturationScore11.Normal||Normal||||||F|||20211209124436|100024000517|123|MEWS
OBX|18|ST|SpO2::SaturationScore11.White||White||||||F|||20211209124436|100024000517|123|MEWS
OBX|19|NM|NIBP::SystolicScore1||0||||||F|||20211209124436|100024000517|123|MEWS
OBX|20|ST|NIBP::SystolicScore1.Normal||Normal||||||F|||20211209124436|100024000517|123|MEWS
OBX|21|ST|NIBP::SystolicScore1.White||White||||||F|||20211209124436|100024000517|123|MEWS
OBX|22|NM|Respiration RateScore3||0||||||F|||20211209124436|100024000517|123|MEWS
OBX|23|ST|Respiration RateScore3.Normal||Normal||||||F|||20211209124436|100024000517|123|MEWS
OBX|24|ST|Respiration RateScore3.White||White||||||F|||20211209124436|100024000517|123|MEWS
OBX|25|NM|TemperatureScore5||0||||||F|||20211209124436|100024000517|123|MEWS
OBX|26|ST|TemperatureScore5.Normal||Normal||||||F|||20211209124436|100024000517|123|MEWS
OBX|27|ST|TemperatureScore5.White||White||||||F|||20211209124436|100024000517|123|MEWS
OBX|28|NM|Pulse RateScore2||0||||||F|||20211209124436|100024000517|123|MEWS
OBX|29|ST|Pulse RateScore2.Normal||Normal||||||F|||20211209124436|100024000517|123|MEWS
OBX|30|ST|Pulse RateScore2.White||White||||||F|||20211209124436|100024000517|123|MEWS
OBX|31|NM|BewustzijnScore11||3||||||F|||20211209124436|100024000517|123|MEWS
OBX|32|ST|BewustzijnScore11.High||High||||||F|||20211209124436|100024000517|123|MEWS
OBX|33|ST|BewustzijnScore11.Red||Red||||||F|||20211209124436|100024000517|123|MEWS
OBX|34|NM|ZuurstoftoedieningScore9||2||||||F|||20211209124436|100024000517|123|MEWS
OBX|35|ST|ZuurstoftoedieningScore9.Medium||Medium||||||F|||20211209124436|100024000517|123|MEWS
OBX|36|ST|ZuurstoftoedieningScore9.Yellow||Yellow||||||F|||20211209124436|100024000517|123|MEWS
OBX|37|NM|MEWS||5||||||F|||20211209124436|100024000517|123|MEWS
OBX|38|ST|MEWS.Low||Low||||||F|||20211209124436|100024000517|123|MEWS
OBX|39|ST|MEWS.Green||Green||||||F|||20211209124436|100024000517|123|MEWS
OBX|40|ST|MEWS.FORMULA||SUM||||||F|||20211209124436|100024000517|123|MEWS

And here is the transformed message:

MSH|^~\&|POWERFLEX|Welch Allyn|HIS||20211209124449||ORU^R01|014AD5C5-5CE6-444C-B010-78055F5FB6E1|P|2.5
PID|||0012345678||TYSON^MIKE
PV1||I|||||||||||||||||0087654321
OBR||||S|||20211209124436|||123|||||||||||||||F
OBX|2|NM|BMI||23.0|||null|||F|||20211209124436|100024000517|123||Manual
OBX|1|NM|0002-F98C^Pijn^MDIL||5|||N|||F|||20211209124436|100024000517|123||Manual
OBX|2|NM|0002-F828^SpRF^MDIL||14|/min||N|||F|||20211209124436|100024000517|123||Manual
OBX|3|NM|0002-4A05^NiBDs^MDIL||120|mmHg||N|||F|||20211209124436|100024000517|123||Manual
OBX|5|NM|0002-4A07^NiBDm^MDIL||0.0|mmHg||null|||F|||20211209124436|100024000517|123||Manual
OBX|6|NM|0002-4BB8^SpO2^MDIL||99|0004-0220^%^MDIL||N|||F|||20211209124436|100024000517|123||Manual
OBX|11|ST|TEMPSITE||None||||||F|||20211209124436|100024000517|123||Manual
OBX|7|NM|0002-FB4B^iTtymp^MDIL||36.9|°C||N|||F|||20211209124436|100024000517|123||Manual
OBX|8|NM|0002-4822^Pols^MDIL||65|/min||N|||F|||20211209124436|100024000517|123||Manual
OBX|9|ST|0002-F989^Bewustzijn^MDIL||Pijn reactie|||null|||F|||20211209124436|100024000517|123
OBX|16|NM|SpO2::SaturationScore11||0||||||F|||20211209124436|100024000517|123|MEWS
OBX|17|ST|SpO2::SaturationScore11.Normal||Normal||||||F|||20211209124436|100024000517|123|MEWS
OBX|18|ST|SpO2::SaturationScore11.White||White||||||F|||20211209124436|100024000517|123|MEWS
OBX|19|NM|NIBP::SystolicScore1||0||||||F|||20211209124436|100024000517|123|MEWS
OBX|20|ST|NIBP::SystolicScore1.Normal||Normal||||||F|||20211209124436|100024000517|123|MEWS
OBX|22|NM|Respiration RateScore3||0||||||F|||20211209124436|100024000517|123|MEWS
OBX|24|ST|Respiration RateScore3.White||White||||||F|||20211209124436|100024000517|123|MEWS
OBX|26|ST|TemperatureScore5.Normal||Normal||||||F|||20211209124436|100024000517|123|MEWS
OBX|28|NM|Pulse RateScore2||0||||||F|||20211209124436|100024000517|123|MEWS
OBX|30|ST|Pulse RateScore2.White||White||||||F|||20211209124436|100024000517|123|MEWS
OBX|32|ST|BewustzijnScore11.High||High||||||F|||20211209124436|100024000517|123|MEWS
OBX|34|NM|ZuurstoftoedieningScore9||2||||||F|||20211209124436|100024000517|123|MEWS
OBX|36|ST|ZuurstoftoedieningScore9.Yellow||Yellow||||||F|||20211209124436|100024000517|123|MEWS
OBX|38|ST|MEWS.Low||Low||||||F|||20211209124436|100024000517|123|MEWS
OBX|40|ST|MEWS.FORMULA||SUM||||||F|||20211209124436|100024000517|123|MEWS

As you can see, the transformed message doesn't filter out the unwanted segments and it screws up the numbering.   I'd appreciate some help to get my code working - thanks in advance!

**Comment by pacmano1** [comment]
Remember when you are deleting segments, you are actually deleting XML.  So you are removing nodes and changing the index.

See https://forums.mirthproject.io/forum/mirth-connect/support/180223-delete-nte-segments-until-finding-matching-string

**Reply by pacmano1** [reply]
Not your exact use case but discusses the concept I am talking about.

**Reply by pacmano1** [reply]
And thanks for posting a thorough example - please do use code highlighting next time.  See https://github.com/nextgenhealthcare/connect/discussions/4824 and https://docs.github.com/en/github/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks

**Reply by Pensive-Potato** [reply]
Thanks very much for the tips, I'll see if I can work this out based on that.  I've edited my original post to highlight the code, thanks for pointing that out.

**Comment by tonygermano** [comment, marked as answer]
Try changing both
```js
delete msg.OBX[counter];
```
to 
```js
delete obx[0]; // delete current obx segment
continue; // start from top of loop with next obx segment
```
You shouldn't need your `counter` variable, and you want to make sure that you don't continue processing the current obx segment if you intended to delete it.

**Reply by Pensive-Potato** [reply]
Brilliant! That did the trick instantly, thank you!!! :1st_place_medal: 

The white List works now, though if I may be so cheeky as to ask a last question...

The white List array itself has been created in the destination transformer; ideally I'd like to get it in the deploy script along with the obx 3.1 and 6.1 formatting.  Is it possible to 'call' it from there?  

Many thanks for your help,

PP
